File: java/business-context/src/main/java/cz/filipklimes/diploma/framework/businessContext/expression/numeric/LessThan.java
package cz.filipklimes.diploma.framework.businessContext.expression.numeric;

import cz.filipklimes.diploma.framework.businessContext.expression.BinaryOperator;
import cz.filipklimes.diploma.framework.businessContext.expression.Expression;
import cz.filipklimes.diploma.framework.businessContext.expression.ExpressionVisitor;
import cz.filipklimes.diploma.framework.businessContext.weaver.BusinessOperationContext;

import java.io.*;
import java.math.BigDecimal;
import java.util.*;

public class LessThan extends BinaryOperator<Boolean, BigDecimal, BigDecimal> implements Serializable
{

    public LessThan(final Expression<BigDecimal> left, final Expression<BigDecimal> right)
    {
        super(left, right);
    }

    @Override
    public Boolean interpret(final BusinessOperationContext context)
    {
        return getLeft().interpret(context).compareTo(getRight().interpret(context)) < 0;
    }

    @Override
    public Map<String, String> getProperties()
    {
        return Collections.emptyMap();
    }

    @Override
    public String getName()
    {
        return "numeric-lt";
    }

    @Override
    public void accept(final ExpressionVisitor visitor)
    {
        visitor.visit(this);
    }

    @Override
    public String toString()
    {
        return String.format("%s < %s", getLeft(), getRight());
    }

}

File: nodejs/business-context/src/expression/numeric/LessThan.js
export default class LessThan {
    constructor(left, right) {
        this.left = left
        this.right = right
    }

    interpret(context) {
        return this.left.interpret(context) < this.right.interpret(context)
    }

    getName() {
        return 'numeric-lt'
    }

    getArguments() {
        return [this.left, this.right]
    }

    getProperties() {
        return {}
    }

    toString() {
        return this.left.toString() + " < " + this.right.toString()
    }

    clone() {
        return new LessThan(
            this.left.clone(),
            this.right.clone(),
        )
    }
}
